pkg/controller/kfservice/kfservice_controller.go

```diff
@@ -108,29 +122,58 @@ func (r *ReconcileService) Reconcile(request reconcile.Request) (reconcile.Resul
 		return reconcile.Result{}, err
 	}
 
-	// TODO(@yuzisun): Define the desired service object
-	service := &knservingv1alpha1.Service{}
-	if err := controllerutil.SetControllerReference(instance, service, r.scheme); err != nil {
+	desiredService, err := resources.CreateKnService(instance)
+	if err != nil {
+		log.Error(err, "Failed to create desired knservice", "name", desiredService.Name)
+		r.Recorder.Eventf(instance, v1.EventTypeWarning, "InternalError", err.Error())
 		return reconcile.Result{}, err
 	}
-
-	found := &knservingv1alpha1.Service{}
-	err = r.Get(context.TODO(), types.NamespacedName{Name: service.Name, Namespace: service.Namespace}, found)
-	if err != nil && errors.IsNotFound(err) {
-		log.Info("Creating service", "namespace", service.Namespace, "name", service.Name)
-		err = r.Create(context.TODO(), service)
-		return reconcile.Result{}, err
-	} else if err != nil {
+	if err := controllerutil.SetControllerReference(instance, desiredService, r.scheme); err != nil {
 		return reconcile.Result{}, err
 	}
 
-	if !reflect.DeepEqual(service.Spec, found.Spec) {
-		found.Spec = service.Spec
-		log.Info("Updating service", "namespace", service.Namespace, "name", service.Name)
-		err = r.Update(context.TODO(), found)
-		if err != nil {
-			return reconcile.Result{}, err
-		}
+	serviceReconciler := knservice.NewServiceReconcile(r.Client)
+
+	knService, err := serviceReconciler.Reconcile(context.TODO(), desiredService)
+	if err != nil {
```

Comment:
combine?

Reply:
Do you mean 

```go
if knService, err := serviceReconciler.Reconcile(context.TODO(), desiredService); err != nil {
```

I think the current way is better since the knService will be used after the if-else block.

Reply:
ye, `knService` is referenced later.